### PR TITLE
FAI-11597: PR Reviews Source + Converter

### PR DESCRIPTION
**Pull request by matiaslcoulougian**
## Description

> Provide description here

## Type of change
- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change

## Related issues

> Fix [#1]() 

## Migration notes

> Describe migration notes if any

## Extra info

> Add any additional information

### CODE REVIEW

File: faros-airbyte-common/resources/github/queries/pull-request-reviews-query.gql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+          hasNextPage
+          endCursor
+        }
+        nodes {
+          ...ReviewFields
+        }
+      }
+    }
+  }
+}
```
Comment (chalenge) [inline faros-airbyte-common/resources/github/queries/pull-request-reviews-query.gql:15]: nit: new line

File: destinations/airbyte-faros-destination/package.json [hunk 2]
```diff
@@ -28,7 +28,7 @@
     "fix": "prettier --write 'src/**/*.ts' 'test/**/*.ts' && npm run lint -- --fix",
     "lint": "eslint 'src/**/*.ts' 'test/**/*.ts'",
     "prepare": "npm run build",
-    "test": "jest --verbose --color",
+    "test": "jest --verbose --color --updateSnapshot",
```
Comment (dbruno21) [inline destinations/airbyte-faros-destination/package.json:31]: remove

File: sources/github-source/package.json [hunk 3]
```diff
@@ -25,7 +25,7 @@
     "fix": "prettier --write 'src/**/*.ts' 'test/**/*.ts' && npm run lint -- --fix",
     "lint": "eslint 'src/**/*.ts' 'test/**/*.ts'",
     "prepare": "npm run build",
-    "test": "jest --verbose --color",
+    "test": "jest --verbose --color --updateSnapshot",
```
Comment (chalenge) [inline sources/github-source/package.json:28]: revert

File: sources/github-source/src/github.ts [hunk 4]
```diff
@@ -67,6 +72,7 @@ export abstract class GitHub {
     private readonly bucketId: number,
     private readonly bucketTotal: number,
     private readonly fetchFiles: boolean,
```
Comment (dbruno21) [inline sources/github-source/src/github.ts:74]: fetchPullRequestFiles

File: sources/github-source/src/github.ts [hunk 5]
```diff
@@ -67,6 +72,7 @@ export abstract class GitHub {
...
+    private readonly fetchReviews: boolean,
```
Comment (dbruno21) [inline sources/github-source/src/github.ts:75]: fetchPullRequestReviews

File: sources/github-source/src/github.ts [hunk 6]
```diff
@@ -175,32 +181,45 @@ export abstract class GitHub {
...
     for await (const res of this.wrapIterable(iter, this.timeout)) {
-      this.logger.info(
-        `Rate limit info si ${JSON.stringify(res['rateLimit'])}`
-      );
       for (const pr of res.repository.pullRequests.nodes) {
         if (cutoffDate && Utils.toDate(pr.updatedAt) <= cutoffDate) {
           break;
         }
         const files = this.fetchFiles ? await this.getFiles(pr, org, repo) : [];
+        const reviews = this.fetchReviews
```
Comment (chalenge) [inline sources/github-source/src/github.ts:189]: let's always return a list, could be empty and not have this other conditional

File: sources/github-source/src/github.ts [hunk 7]
```diff
@@ -251,6 +270,52 @@ export abstract class GitHub {
...
 
+  private async getReviews(
+    pr: PullRequestNode,
+    org: string,
+    repo: string
+  ): Promise<PullRequestReview[]> {
+    let reviews = pr.reviews.nodes;
+    const {hasNextPage, endCursor} = pr.reviews.pageInfo;
+    if (hasNextPage) {
+      const remainingReviews = await this.getPullRequestReviews(
```
Comment (chalenge) [inline sources/github-source/src/github.ts:281]: what did do for PR Files? Kinda okay with graphQL here but let's have it consistent for other nested nodes

File: sources/github-source/src/github.ts [hunk 8]
```diff
@@ -251,6 +270,52 @@ export abstract class GitHub {
...
+        org,
+        repo,
+        pr.number,
+        endCursor
+      );
+      reviews = reviews.concat(remainingReviews);
+    }
+    return reviews;
```
Comment (chalenge) [inline sources/github-source/src/github.ts:289]: if we think this maybe nullish then `return reviews ?? []` so we do not need to handle it in main code

File: sources/jira-source/package.json [hunk 9]
```diff
@@ -24,7 +24,7 @@
     "fix": "prettier --write 'src/**/*.ts' 'test/**/*.ts' && npm run lint -- --fix",
     "lint": "eslint 'src/**/*.ts' 'test/**/*.ts'",
     "prepare": "npm run build",
-    "test": "jest --verbose --color",
+    "test": "jest --verbose --color --updateSnapshot",
```
Comment (chalenge) [inline sources/jira-source/package.json:27]: revert

**Review by chalenge — CHANGES_REQUESTED** [review]
Add the permissions needed for this stream

File: sources/github-source/src/github.ts (hunk 7 above)
Comment (matiaslcoulougian) [reply sources/github-source/src/github.ts:281]: For files we considered best using the rest API, cause the request returned all the fields we needed and we are avoiding getting nearer the GraphQL API rate limit. In this case opted by a graphql query cause the endpoint to list reviews by pr does not include comments count, meaning we would need to do +1 request to get this data and obtain the complete object.

File: sources/github-source/src/github.ts (hunk 8 above)
Comment (matiaslcoulougian) [reply sources/github-source/src/github.ts:300]: Should not be nullish if they are no reviews. API returns an empty array

File: destinations/airbyte-faros-destination/src/converters/github/faros_pull_requests.ts [hunk 10]
```diff
@@ -111,17 +111,29 @@ export class FarosPullRequests extends GitHubConverter {
...
+        model: 'vcs_PullRequestFile',
+        record: {
+          pullRequest: prKey,
+          file: {uid: file.path, repository: repoKey},
+          additions: file.additions,
+          deletions: file.deletions,
+        },
+      })),
+      ...pr.reviews.map((review) => ({
+        model: 'vcs_PullRequestReview',
```
Comment (dbruno21) [inline destinations/airbyte-faros-destination/src/converters/github/faros_pull_requests.ts:124]: Add to destination models

**Comment by dbruno21** [comment]
@matiaslcoulougian What about vcs_PullRequestReviewRequest? Can create ticket and do on another PR

File: destinations/airbyte-faros-destination/src/converters/github/faros_pull_requests.ts [hunk 11]
```diff
@@ -111,17 +111,29 @@ export class FarosPullRequests extends GitHubConverter {
...
+        record: {
+          number: review.databaseId,
+          uid: review.databaseId.toString(),
+          htmlUrl: review.url,
+          pullRequest: prKey,
+          reviewer: review.author
```
Comment (dbruno21) [inline destinations/airbyte-faros-destination/src/converters/github/faros_pull_requests.ts:130]: Each review author should be collected as we do for PR author
https://github.com/faros-ai/airbyte-connectors/blob/992b6d4a6ea0355abdbc127c85f8811ace079df1/destinations/airbyte-faros-destination/src/converters/github/faros_pull_requests.ts#L39

**Comment by matiaslcoulougian** [comment]
> vcs_PullRequestReviewRequest

Missed it from the models included in this ticket. I will add in following PR 👍

File: sources/github-source/src/github.ts (hunk 8 above)
Comment (chalenge) [reply sources/github-source/src/github.ts:300]: I had earlier seen us handling nullish condition in the main code

**Review by chalenge — APPROVED** [review]
lgtm

File: sources/github-source/src/github.ts (hunk 8 above)
Comment (matiaslcoulougian) [reply sources/github-source/src/github.ts:300]: Maybe the condition was on the `fetchPullRequestReviews` flag. If that flag is off, we remove the reviews fragment for the query so pr.reviews will be null. But now we are directly checking for that `fetchPullRequestReviews` value: https://github.com/faros-ai/airbyte-connectors/pull/1600/files#diff-c3ec475eeb0ddb699ccf89b07eb40f24741bc8974d505f77035e1a8d27f4e563R286

**Pull request merged**